enhancements/api-review/azure_user_defined_tags.md

```diff
@@ -0,0 +1,317 @@
+reviewers:
+  - "@patrickdillon" ## reviewer for installer component
+  - "@JoelSpeed" ## reviewer for api and machine-api-provider-azure components
+  - "@dmage" ## reviewer for cluster-image-registry-operator component
+  - "@Miciah" ## reviewer for cluster-ingress-operator component
+  - "@akhil-rane" ## reviewer for cloud-credential-operator component
+  - "@trozet" ## reviewer for cloud-network-config-controller component
+approvers:
+  - "@jerpeter1" ## approver for CFE
+api-approvers:
+  - "@JoelSpeed" ## approver for api component
+creation-date: 2022-07-12
+last-updated: 2022-07-12
+tracking-link:
+  - https://issues.redhat.com/browse/OCPPLAN-8155
+  - https://issues.redhat.com/browse/CORS-2249
+see-also:
+  - "enhancements/api-review/custom-tags-aws.md"
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Apply user defined tags to all Azure resources created by OpenShift
+
+## Summary
+
+This enhancement describes the proposal to allow an administrator of Openshift to 
+have the ability to apply user defined tags to those resources created by Openshift 
+in Azure.
+
+## Motivation
+
+Motivations include but are not limited to:
+
+- Allow admin, compliance, and security teams to keep track of assets and objects 
+  created by OpenShift in Azure.
+
+### User Stories
+
+- As an openshift administrator, I want to have tags added to all resources created 
+  in Azure by Openshift, so that I can restrict access granted to an OpenShift specific account
+
+### Goals
+
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
+  can configure allowed number of user-defined tags in the Openshift installer generated
+  install config, which is referred and applied by the installer and the in-cluster operators
+  on the the Azure resources during cluster creation.
+- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
+  to create an object and to apply tags post cluster creation.
+
+### Non-Goals
+
+- Management(update/delete) of resource tags post creation of cluster is out of scope.
+
+## Proposal
+
+A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
+resource created by Openshift to enable administrator to differentiate the resources
+created for Openshift cluster. An administrator is not allowed to add or modify the tag 
+having the prefix `kubernetes.io` or `openshift.io` in the name. The same tag can be 
+found applied to other cloud platform resources which supports tagging for ex: AWS.
+
+New `userTags` field will be added to `platform.azure` of install-config for the user 
+to define the tags to be added to the resources created by installer and in-cluster operators.
+
+If `platform.azure.userTags` of install-config has any tag defined same will be added 
+to all the azure resources created by Openshift except, when the tag validation fail
+to meet any of the below conditions
+1. A tag name can have a maximum of 128 characters.
```

Comment:
Do we have any limit on the name of clusters? Could the combination of a long cluster name and the `kubernetes.io/cluster/` prefix cause us to blow this limit? And if so, how do we handle that?

Reply:
Cluster name [generated](https://github.com/openshift/api/blob/28228c0f01ab20e1d49d8369157fb171ad53e102/config/v1/types_infrastructure.go#L52-L55) by appending a five characters length random string(+1 for `-` used as delimiter) to the user inputted name is restricted to 27 characters, which is within the limit and is not a problem I think.
/cc @patrickdillon Could you please confirm, if my understanding is correct?

Reply:
If we can verify that's correct, it might be worth noting that somewhere within the enhancement to prevent others asking the same question

Reply:
I did verify deploying a cluster on Azure. The first catch is when user inputs the cluster name during install-config creation an error is thrown if name has more than 54 characters. And later during the generation of InfraID, if the provided cluster name has more than 21 characters, it is trimmed to 21 characters and a 5 character random string is appended with `-` as a delimiter. The maximum length of a default tag will come to 49 characters.

Reply:
Ok, please make sure this gets noted in the enhancement somewhere, useful breadcrumbs

Reply:
Content [added](https://github.com/openshift/enhancements/pull/1191/commits/5cd152284b32fb3ca953b55b031a7c52c9c9e078) to the enhancement.